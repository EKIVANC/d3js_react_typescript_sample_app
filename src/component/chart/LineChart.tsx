import React, { useEffect, useRef } from 'react';
import * as d3 from 'd3';
import IChartData from '../../interface/IChartData';

const plotMargins = {
  top: 30,
  bottom: 30,
  left: 30,
  right: 30,
};

const width = 960;
const height = 480;

type LineChartProps = {
  data: IChartData[];
};

export const LineChart = ({ data }: LineChartProps): JSX.Element => {
  const container = useRef(null);

  // work only component !
  useEffect(() => {
    if (!data) {
      return;
    }
    const canvas = d3
      .select(container.current)
      .append('svg')
      .attr('width', width)
      .attr('height', height);

    const xMin =
      d3.min(data, (d) => {
        return new Date(d.x);
      }) || new Date();
    const xMmax =
      d3.max(
        data,
        (d): Date => {
          return new Date(d.x);
        }
      ) || new Date();
    const yMin =
      d3.min(data, (d) => {
        return d.y;
      }) || 0;
    const yMmax =
      d3.max(data, (d) => {
        return d.y;
      }) || 0;

    const xScale = d3
      .scaleTime()
      .domain([xMin, xMmax])
      .range([plotMargins.left, width - plotMargins.right])
      .nice();
    const yScale = d3
      .scaleLinear()
      .domain([yMmax, yMin])
      .range([plotMargins.top, height - plotMargins.bottom]);

    const xAxis = d3.axisBottom(xScale);
    const yAxis = d3.axisRight(yScale);
    //create x Axis
    canvas
      .append('g')
      .attr('transform', 'translate(0,' + (height - plotMargins.bottom) + ')')
      .call(xAxis);
    //create y Axis
    canvas
      .append('g')
      .attr('transform', 'translate( ' + plotMargins.left + '  ,0)')
      .call(yAxis);

    // define the line generator
    const lineGenerator = d3
      .line<IChartData>()
      .x((d) => {
        return xScale(d['x']);
      })
      .y((d) => {
        return yScale(d['y']);
      })
      .curve(d3.curveBasis);

    const line = lineGenerator(data) || '';

    canvas
      .append('path')
      .attr('d', line)
      .attr('stroke', 'green')
      .attr('stroke-width', 2)
      .attr('fill', 'none');
  }, []);

  return <div className="line-chart" ref={container} />;
};
