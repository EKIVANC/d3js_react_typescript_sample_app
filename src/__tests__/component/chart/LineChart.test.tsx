import React from "react";
import { create } from "react-test-renderer";
import {LineChart} from "../../../component/chart/LineChart";

describe("LineChart component", () => {
    test("Matches the snapshot", () => {
        const mockData = [{
            x: new Date(),
            y: 10
        }]
        const lineChart = create(<LineChart data={mockData} />);
        expect(lineChart.toJSON()).toMatchSnapshot();
    });
});
