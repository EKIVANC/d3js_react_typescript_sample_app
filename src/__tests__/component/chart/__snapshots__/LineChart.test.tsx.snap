// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart component Matches the snapshot 1`] = `
<div
  className="line-chart"
/>
`;
